text/0000-let-expression.md

```diff
@@ -0,0 +1,996 @@
+RFC and similar RFCs in the future won't be happen and their task will be taken with this consistent syntax.
+
+
+### Compare expressive power of this RFC with let-else
+
+```rust
+// simple let else
+let Some(x) = y else {
+    return Err("fail");
+};
+
+// with let expression
+let Some(x) = y || {
+    return Err("fail");
+};
+
+// or even better
+let Some(x) = y || return Err("fail");
+
+// let else else future possiblity
+let Some(x) = a else b else c else { return; };
+
+// with let expression
+let Some(x) = a
+|| let Some(x) = b
+|| let Foo(x) = bar
+|| return;
+
+// duplicate else block of consecutive let-else
+let Some(Foo(x)) = bar else {
+    panic!("a very long message which needs to change every day");
+};
+let Some((y, z)) = baz(x) else {
+    panic!("a very long message which needs to change every day");
+};
+
+// with let expression
+let Some(Foo(x)) = bar
+&& let Some((y, z)) = baz(x)
+|| panic!("a very long message which needs to change every day");
+```
+
+### `else` vs `||`
+
+Some people argue that `else` is a better choice and `||` doesn't read very well. But in fact using
+short circuit operators in this way is a wellknown pattern in general, and it is popular in bash scripting. In
+standard ML, short circuit operators are called `OrElse` and `AndAlso` which shows that this similarity is known
+and `else` in let-else is more like `OrElse` rather than `else` in if-else, so `||` is a good choice.
+
+As a benefit for `||` over `else`, `||` is already in the language and working for
+normal bools ([playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=13b6c01bca742a083b0f48fee07b21d3)).
+but `is_prime(x) else { continue };` isn't and won't be valid syntax in rust. So this RFC need much less changes
+in the language.
+
+For consistency with bool expressions, we should allow something like `let Some(x) = y || panic!("fail")`. For
+normal bools it is considered a bad practice, (though [not everyone is agree](https://github.com/rust-lang/rust/issues/69466#issuecomment-591097014)
+with it) and explicit `if` is preferred. In let expressions, explicit `if` is not an option due to difference in bindings,
+and demand for such construct is real (The presence of let-else is a sign of this), so why we should consider a
+pattern that exists in current rust and many other languages a bad practice, and then invent a new construct
+that do exactly the same work and read almost the same (let-else vs let-orelse) and consider it a good practice?
+
+## A subset of this RFC
+Not all things introduced here are useful and some of them are because consistency and completeness. We can make
+some subsets of this RFC hard errors and make them future possiblity. Subsets that are candidate of removing
+are:
+* The `!` operator:
+we can remove `!` operator so there would be no NB and whole thing becomes simpler to understand. But there may
```

Comment:
Let's expand this acronym, because it's not obvious from context
```suggestion
we can remove `!` operator so there would be no negative bindings, and whole thing becomes simpler to understand. But there may
```

Reply:
Ah, I missed it, whole part need to be removed. In old (!) days, NB was a thing :) Thanks.